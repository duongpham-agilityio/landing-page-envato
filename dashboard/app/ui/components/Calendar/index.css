/* Month view */
.rbc-month-view {
  border-width: 0;
}

.rbc-month-header .rbc-header {
  font-weight: 600;
  font-size: 20px;
  line-height: 28px;
  color: #22c55e;
  border: none !important;
}

.rbc-month-header {
  padding-top: 20px;
  border-width: 0;
  border-top: 1px solid light-dark(#e4f2ff, #2a313c);
  margin-bottom: 30px;
}

.rbc-month-header .rbc-header:first-child {
  color: #ff4747;
}

.rbc-month-row .rbc-date-cell {
  font-size: 18px;
  line-height: 18px;
  color: light-dark(#2f3032, #ffffff);
  background: transparent;
  border: none !important;
  text-align: center;
}

.rbc-month-row .rbc-date-cell:first-child,
.rbc-date-cell:last-child {
  color: #ff4747;
}

.rbc-month-row .rbc-off-range.rbc-date-cell {
  color: #c0c0c0;
}

.rbc-month-row .rbc-off-range.rbc-date-cell {
  color: #c0c0c0;
}

.rbc-month-row .rbc-off-range.rbc-date-cell:first-child,
.rbc-off-range.rbc-date-cell:last-child {
  color: #ffc7c7;
}

.rbc-month-row .rbc-now.rbc-date-cell {
  color: #22c55e;
}

.rbc-month-row .rbc-off-range-bg {
  background-color: transparent;
}

.rbc-month-row .rbc-button-link {
  font-weight: 500;
}

.rbc-month-row .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: light-dark(#2f3032, #ffffff);
}

.rbc-month-row .rbc-off-range .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: #c0c0c0;
}

.rbc-month-row .rbc-date-cell:first-child .rbc-button-link:hover,
.rbc-month-row .rbc-date-cell:last-child .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: #ff4747;
}

.rbc-month-row .rbc-off-range:first-child .rbc-button-link:hover,
.rbc-month-row .rbc-off-range:last-child .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: #ffc7c7;
}

.rbc-month-row .rbc-off-range .rbc-button-link:hover {
  text-decoration: underline;
  text-decoration-color: #c0c0c0;
}

.rbc-month-row .rbc-today {
  background-color: transparent;
}

.rbc-month-row,
.rbc-day-bg {
  border: none !important;
}

.rbc-month-view .rbc-event {
  font-size: 14px;
  line-height: 24px;
  font-weight: 700;
  color: light-dark(#000, #fff);
  background-color: transparent;
}

.rbc-month-view .rbc-event:hover {
  background-color: light-dark(#f5f5f5, #1a1b20);
}

.rbc-month-view .rbc-event-content::before {
  content: '';
  display: inline-block;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  background-color: #f5ba47;
  margin-right: 6px;
}

/* Week view */
.rbc-events-container .rbc-event {
  background-color: #f5ba47;
  border-color: #ffffff;
}

.rbc-time-view .rbc-event .rbc-event-label {
  font-size: 13px;
}

.rbc-time-view .rbc-event .rbc-event-content {
  font-size: 15px;
}

/* Calendar */
.rbc-calendar {
  min-height: 530px;
}

@media screen and (min-width: 768px) {
  .rbc-calendar {
    min-height: 700px;
  }
}

@media screen and (min-width: 992px) {
  .rbc-calendar {
    min-height: 870px;
  }
}
