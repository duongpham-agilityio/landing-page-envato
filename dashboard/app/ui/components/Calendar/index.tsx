'use client';

// Libs
import { useCallback, useState } from 'react';
import {
  Calendar as BigCalendar,
  momentLocalizer,
  CalendarProps,
  Views,
} from 'react-big-calendar';
import moment from 'moment';
import 'react-big-calendar/lib/css/react-big-calendar.css';

// Components
import CustomToolbar from './CustomToolBar';

type ViewType = 'month' | 'week' | 'work_week' | 'day' | 'agenda';

const localizer = momentLocalizer(moment);

const Calendar = (props: Omit<CalendarProps, 'localizer'>) => {
  const [date, setDate] = useState(new Date());
  const [view, setView] = useState<ViewType>(Views.MONTH);

  const onNavigate = useCallback(
    (newDate: Date) => setDate(newDate),
    [setDate],
  );

  const onView = useCallback(
    (newView: ViewType) => setView(newView),
    [setView],
  );

  return (
    <BigCalendar
      {...props}
      localizer={localizer}
      startAccessor="start"
      endAccessor="end"
      style={{ height: '50vh' }}
      date={date}
      onNavigate={onNavigate}
      defaultView={Views.MONTH}
      view={view}
      views={[Views.MONTH, Views.WEEK, Views.DAY]}
      onView={onView}
      components={{ toolbar: CustomToolbar }}
    />
  );
};

export default Calendar;
