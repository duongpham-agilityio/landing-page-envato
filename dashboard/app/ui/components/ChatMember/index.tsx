import { memo } from 'react';
import {
  Box,
  Flex,
  Text,
  Avatar,
  AvatarBadge,
  useColorModeValue,
  useBreakpointValue,
} from '@chakra-ui/react';

// Themes
import { colors } from '@/ui/themes/bases/colors';

// Utils
import { getStatusColor } from '@/lib/utils';

// Constants
import { IMAGES } from '@/lib/constants';

export type Props = {
  avatar?: string;
  name?: string;
  localeTime?: string;
  icon?: React.ReactNode;
  statusColor?: string;
  onClick?: () => void;
  lastMessages?: string;
};

const ChatMember = ({
  avatar = IMAGES.CHAT_USER_AVATAR.url,
  name = 'John Doe',
  lastMessages,
  localeTime,
  icon,
  statusColor = '',
  onClick = () => {},
}: Props) => {
  const colorFill = useColorModeValue(
    colors.secondary[200],
    colors.secondary[600],
  );

  const isMobile = useBreakpointValue({ base: true, lg: false });

  return isMobile ? (
    <Box
      cursor="pointer"
      _hover={{ bg: colorFill }}
      onClick={onClick}
      borderRadius="lg"
    >
      <Flex justify="space-between" p={3.5}>
        <Flex gap={3} borderRadius="50%" border="1px solid">
          <Avatar src={avatar} borderRadius="50%">
            <AvatarBadge boxSize={4} bg={getStatusColor(statusColor)} top={7} />
          </Avatar>
        </Flex>

        <Flex direction="column" alignItems="center">
          <Text>{localeTime}</Text>
          {icon}
        </Flex>
      </Flex>
    </Box>
  ) : (
    <Box
      cursor="pointer"
      _hover={{ bg: colorFill }}
      onClick={onClick}
      borderRadius="lg"
    >
      <Flex justify="space-between" p={3.5}>
        <Flex gap={3}>
          <Avatar
            src={avatar}
            borderRadius="50%"
            border="1px solid"
            borderColor="border.tertiary"
          >
            <AvatarBadge boxSize={4} bg={getStatusColor(statusColor)} top={7} />
          </Avatar>
          <Flex flexDirection="column">
            <Box mr={6}>
              <Text fontSize="24px" fontWeight="bold">
                {name}
              </Text>
            </Box>
            <Text color="primary.300">{lastMessages}</Text>
          </Flex>
        </Flex>

        <Flex direction="column" alignItems="center">
          <Text>{localeTime}</Text>
          {icon}
        </Flex>
      </Flex>
    </Box>
  );
};

const ChatMemberMemorized = memo(ChatMember);

export default ChatMemberMemorized;
