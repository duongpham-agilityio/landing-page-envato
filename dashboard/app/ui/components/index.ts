export { default as InputField } from './common/InputField';
export { default as FallbackErrorBoundary } from './FallbackErrorBoundary';
export { default as Navigation } from './common/Navigation';
export { default as Menu } from './common/Menu';
export { default as Logo } from './common/Logo';
export { default as ExpandSidebar } from './ExpandSidebar';
export { default as MiniSidebar } from './MiniSidebar';
export * from './Icons';
export * from './Images';
export { default as Avatar } from './common/Avatar';
export { default as Benefit } from './common/Benefit';
export { default as Divider } from './common/Divider';
export { default as IconButton } from './common/IconButton';
export { default as Dropdown } from './common/Dropdown';
export { default as Select } from './common/Select';
export { default as Pagination } from './common/Pagination';
export { default as Button } from './common/Button';
export { default as Table } from './common/Table';
export { default as CustomerNameCell } from './common/Table/Body/CustomerNameCell';
export { default as ActionCell } from './common/Table/Body/ActionCell';
export { default as UserInfoCell } from './UsersTable/Cell/UserInfoCell';
export { default as StatusCell } from './UsersTable/Cell/StatusCell';
export { default as HeadCell } from './common/Table/HeadCell';
export { default as Selector } from './TransactionTable/Selector';
export { default as TransactionTable } from './TransactionTable';
export { default as SearchBar } from './common/SearchBar';
export { default as SwitchTheme } from './SwitchTheme';
export { default as CardPayment } from './CardPayment';
export { default as BoxChat } from './BoxChat';
export { default as TotalStatisticCard } from './TotalStatisticCard';
export { default as Fetching } from './Skeleton/Fetching';
export { default as TotalStatisticListSkeleton } from './Skeleton/TotalStatisticListSkeleton';
export { default as TotalStatisticList } from './TotalStatisticList';
export { default as RevenueFlow } from './RevenueFlow';
export { default as Efficiency } from './Efficiency';
export { default as TotalBalance } from './TotalBalance';
export { default as OverallBalance } from './OverallBalance';
export { default as UserCard } from './UserCard';
export { default as ItemSideBarSetting } from './ItemSideBarSetting';
export { default as FaqItem } from './FaqItem';
export { default as UpdateProfile } from './Profile';
export { default as AuthForm } from './AuthForm';
export { default as Modal } from './common/Modal';
export { default as FallbackImage } from './FallbackImage';
export { default as ChatMember } from './ChatMember';
export { default as Conversation } from './Conversation';
export { default as PinCode } from './common/PinCode';
