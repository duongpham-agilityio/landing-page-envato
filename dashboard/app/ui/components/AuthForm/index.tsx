'use client';

import { ChangeEvent, memo, useCallback } from 'react';
import Link from 'next/link';
import { useRouter } from 'next/navigation';
import { ViewOffIcon, ViewIcon } from '@chakra-ui/icons';
import {
  Button,
  HStack,
  Text,
  VStack,
  Checkbox,
  useDisclosure,
  Box,
  Link as ChakraLink,
  Flex,
  FormErrorMessage,
  Input,
} from '@chakra-ui/react';

import { Controller, SubmitHandler } from 'react-hook-form';

// Hooks
import { useForm, useAuth } from '@/lib/hooks';

// Constants
import { ROUTES, AUTH_SCHEMA, ERROR_MESSAGES } from '@/lib/constants';

// Components
import { InputField } from '@/ui/components';

// Utils
import { validatePassword } from '@/lib/utils';

// Types
import { TUserDetail } from '@/lib/interfaces';
import { getMessaging } from 'firebase/messaging';

// Layouts
import { AuthFooter } from '@/ui/layouts';

// firebase
import { app, requestForToken } from '@/firebase';

type TAuthForm = Omit<TUserDetail, 'id' | 'createdAt'> & {
  confirmPassword: string;
  isAcceptPrivacyPolicy: boolean;
  isRemember: false;
};

interface AuthFormProps {
  isRegister?: boolean;
}

const AuthFormComponent = ({
  isRegister = false,
}: AuthFormProps): JSX.Element => {
  const { signIn, signUp } = useAuth();
  const router = useRouter();

  // Control form
  const {
    control,
    formState: {
      errors: { root },
      dirtyFields,
      defaultValues,
      isSubmitting,
    },
    handleSubmit,
    setError,
    clearErrors,
  } = useForm<TAuthForm>({
    defaultValues: {
      email: '',
      password: '',

      ...(!isRegister && {
        isRemember: false,
      }),
      ...(isRegister && {
        firstName: '',
        lastName: '',
        confirmPassword: '',
        isAcceptPrivacyPolicy: false,
      }),
    },
  });

  // Control show/hide password
  const { isOpen: isShowPassword, onToggle: onShowPassword } = useDisclosure();
  const { isOpen: isShowConfirmPassword, onToggle: onShowConfirmPassword } =
    useDisclosure();

  const messaging = typeof window !== 'undefined' ? getMessaging(app) : null;

  const isDisabledSubmitBtn: boolean = (() => {
    const getLength = (object: object): number => Object.keys(object).length;

    const isFillAllFields: boolean = isRegister
      ? getLength(dirtyFields) === getLength(defaultValues ?? {})
      : getLength(dirtyFields) === getLength(defaultValues ?? {}) - 1;

    return isSubmitting || !isFillAllFields;
  })();

  const renderPasswordIcon = useCallback(
    (isCorrect: boolean, callback: typeof onShowPassword): JSX.Element => {
      const Icon = isCorrect ? ViewIcon : ViewOffIcon;

      return (
        <Icon
          color="gray.400"
          w="25px"
          h="25px"
          cursor="pointer"
          onClick={callback}
        />
      );
    },
    [],
  );

  const handleLogin: SubmitHandler<TAuthForm> = useCallback(
    async (data) => {
      try {
        const { email, password, isRemember } = data;
        const fcmToken =
          (messaging && (await requestForToken(messaging))) || '';

        await signIn({ email, password, fcmToken }, isRemember);

        router.push(ROUTES.ROOT);
      } catch (error) {
        const { message } = error as Error;

        setError('root', { type: 'custom', message });
      }
    },
    [messaging, router, setError, signIn],
  );

  const handleRegister: SubmitHandler<TAuthForm> = useCallback(
    async (data) => {
      const {
        isAcceptPrivacyPolicy: _isAcceptPrivacyPolicy,
        confirmPassword: _confirmPassword,
        ...fieldValues
      } = data;
      try {
        const fcmToken =
          (messaging && (await requestForToken(messaging))) || '';
        const { errors } = await signUp({ ...fieldValues, fcmToken });
        if (errors) {
          return Object.entries(errors).forEach(([key, value]) =>
            setError(key as keyof typeof data, {
              type: 'custom',
              message: value,
            }),
          );
        }
        router.push(ROUTES.ROOT);
      } catch (error) {
        setError('root', {
          type: 'custom',
          message: ERROR_MESSAGES.SOMETHING_ERROR,
        });
      }
    },
    [messaging, router, setError, signUp],
  );

  const handleClearErrorMessage = useCallback(
    (
      field: keyof TAuthForm,
      isError: boolean,
      onChange: (value: string) => void,
    ) =>
      (data: string) => {
        if (isError) {
          clearErrors(field);
        }

        onChange(data);
      },
    [clearErrors],
  );

  const handleClearRootError = useCallback(
    () => clearErrors('root'),
    [clearErrors],
  );

  return (
    <Box
      id={!isRegister ? 'login-form' : 'register-form'}
      as="form"
      onSubmit={handleSubmit(!isRegister ? handleLogin : handleRegister)}
      w={{
        base: '100%',
        sm: 425,
        md: 460,
      }}
      margin="auto"
      pb={16}
      sx={{
        boxSizing: {
          base: 'border-box',
          md: 'unset',
        },
      }}
    >
      <VStack gap={6}>
        {isRegister && (
          <HStack
            gap={{
              base: 6,
              md: 10,
            }}
            w="100%"
            flexDirection={{
              base: 'column',
              md: 'row',
            }}
          >
            <Controller
              control={control}
              rules={AUTH_SCHEMA.FIRST_NAME}
              name="firstName"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  variant="authentication"
                  placeholder="First name"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={handleClearErrorMessage(
                    'firstName',
                    !!error,
                    field.onChange,
                  )}
                  aria-label="first-name"
                />
              )}
            />
            <Controller
              control={control}
              rules={AUTH_SCHEMA.LAST_NAME}
              name="lastName"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  variant="authentication"
                  placeholder="Last name"
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={handleClearErrorMessage(
                    'lastName',
                    !!error,
                    field.onChange,
                  )}
                  aria-label="last-name"
                />
              )}
            />
          </HStack>
        )}
        <Controller
          rules={AUTH_SCHEMA.EMAIL}
          control={control}
          name="email"
          render={({ field: { value, onChange }, fieldState: { error } }) => {
            const handleChange = (event: { target: { value: string } }) => {
              const value: string = event.target.value;
              const sanitizedValue = value.trim();

              onChange(sanitizedValue);
            };

            return (
              <>
                <Input
                  variant="authentication"
                  placeholder="Username or email"
                  isDisabled={isSubmitting}
                  value={value}
                  name="username"
                  onChange={handleChange}
                  onBlur={handleClearRootError}
                />
                {!!error?.message && (
                  <FormErrorMessage>{error?.message}</FormErrorMessage>
                )}
              </>
            );
          }}
        />

        <Controller
          rules={{ validate: validatePassword }}
          control={control}
          name="password"
          render={({ field, fieldState: { error } }) => (
            <InputField
              type={isShowPassword ? 'text' : 'password'}
              variant="authentication"
              placeholder="Password"
              rightIcon={renderPasswordIcon(isShowPassword, onShowPassword)}
              isError={!!error?.message}
              errorMessages={error?.message}
              isDisabled={isSubmitting}
              {...field}
              onBlur={handleClearRootError}
            />
          )}
        />

        {/* Helpers */}
        {!isRegister && (
          <HStack justifyContent="space-between" w="100%" mt={6}>
            <Controller
              control={control}
              name="isRemember"
              render={({ field: { value, onChange } }) => (
                <Checkbox
                  aria-label="remember"
                  variant="round"
                  isChecked={value}
                  onChange={(e: ChangeEvent<HTMLInputElement>) =>
                    onChange(e.target.checked)
                  }
                  position="relative"
                >
                  <Text fontWeight="semibold">Remember me</Text>
                </Checkbox>
              )}
            />
            <Text
              as={Link}
              href={`/${ROUTES.FORGOT_PASSWORD}`}
              aria-label="forgot password"
              color="primary.500"
              fontWeight="semibold"
              textTransform="capitalize"
              textDecoration="underline"
            >
              forgot password?
            </Text>
          </HStack>
        )}

        {isRegister && (
          <>
            <Controller
              control={control}
              rules={AUTH_SCHEMA.CONFIRM_PASSWORD}
              name="confirmPassword"
              render={({ field, fieldState: { error } }) => (
                <InputField
                  type={isShowConfirmPassword ? 'text' : 'password'}
                  variant="authentication"
                  placeholder="Confirm password"
                  rightIcon={renderPasswordIcon(
                    isShowConfirmPassword,
                    onShowConfirmPassword,
                  )}
                  {...field}
                  isError={!!error}
                  errorMessages={error?.message}
                  isDisabled={isSubmitting}
                  onChange={handleClearErrorMessage(
                    'confirmPassword',
                    !!error,
                    field.onChange,
                  )}
                />
              )}
            />

            <Flex gap={3}>
              <Controller
                control={control}
                rules={AUTH_SCHEMA.AGREE_POLICY}
                name="isAcceptPrivacyPolicy"
                render={({
                  field: { value, onChange },
                  fieldState: { error },
                }) => (
                  <Checkbox
                    size="md"
                    colorScheme="green"
                    isChecked={value}
                    isDisabled={isSubmitting}
                    onChange={(e: ChangeEvent<HTMLInputElement>) =>
                      onChange(e.target.checked)
                    }
                    {...(error && {
                      sx: {
                        '> span': {
                          borderColor: 'danger.400',
                        },
                      },
                    })}
                  >
                    <Text color="text.secondary" fontSize="md" flex={1}>
                      By creating an account, you&apos;re agreeing to our {''}
                      <ChakraLink
                        href="#"
                        color="text.primary"
                        cursor="pointer"
                      >
                        Privacy Policy
                      </ChakraLink>
                      , and {''}
                      <ChakraLink
                        href="#"
                        color="text.primary"
                        cursor="pointer"
                      >
                        Electronics Communication Policy.
                      </ChakraLink>
                    </Text>
                  </Checkbox>
                )}
              />
            </Flex>
          </>
        )}
      </VStack>

      {/* Show API error */}
      <Box mb={7}>
        <Text color="red" textAlign="center" py={2} h={10}>
          {root?.message}
        </Text>
        <Button
          type="submit"
          role="button"
          aria-label={!isRegister ? 'Sign In' : 'Sign Up'}
          textTransform="capitalize"
          form={!isRegister ? 'login-form' : 'register-form'}
          isDisabled={isDisabledSubmitBtn}
        >
          {!isRegister ? 'Sign In' : 'Sign Up'}
        </Button>
      </Box>

      <Flex justifyContent="center">
        <Text
          fontWeight="medium"
          textAlign="center"
          dangerouslySetInnerHTML={{
            __html: !isRegister
              ? 'Don&apos;t have an account?'
              : 'Already have an account?',
          }}
        />
        <Text
          as={Link}
          href={!isRegister ? `/${ROUTES.REGISTER}` : `/${ROUTES.LOGIN}`}
          aria-label={!isRegister ? 'sign up' : 'sign in'}
          fontWeight="semibold"
          textDecoration="underline"
          ml={2}
        >
          {!isRegister ? 'Sign Up' : 'Sign In'}
        </Text>
      </Flex>
      <AuthFooter />
    </Box>
  );
};

const AuthForm = memo(AuthFormComponent);

export default AuthForm;
