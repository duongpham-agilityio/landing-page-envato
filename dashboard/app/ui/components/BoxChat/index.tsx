'use client';

import { memo, useCallback, useEffect, useRef, useState } from 'react';
import { Box, Heading, Flex } from '@chakra-ui/react';

// Components
import Message from './Message';

// Components
import { InputSendMessages } from '@/ui/components';

// Interface
import { TMessages } from '@/lib/interfaces';

// Stores
import { authStore } from '@/lib/stores';

// Hooks
import { getCurrentUser } from '@/lib/hooks';

// Firebase
import {
  doc,
  getDoc,
  onSnapshot,
  serverTimestamp,
  setDoc,
  updateDoc,
} from 'firebase/firestore';
import { db } from '@/lib/utils';
import { FIREBASE_CHAT, USER_CHATS_FIELD } from '@/lib/constants';

const initialUserChat = {
  roomChatId: '',
  userId: '',
  adminId: '',
  avatarUrl: '',
  avatarAdminUrl: '',
  displayName: '',
};

const BoxChatComponent = () => {
  const [messages, setMessages] = useState<TMessages[]>([]);
  const user = authStore((state) => state.user);
  const [userChat, setUserChat] = useState(initialUserChat);
  const boxRef = useRef<HTMLDivElement | null>(null);

  const createChatRoom = useCallback(async () => {
    // Get user data
    const usersData = await getCurrentUser(user);

    if (usersData) {
      await setDoc(doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId), {
        messages: [],
      });

      await updateDoc(doc(db, FIREBASE_CHAT.USER_CHATS, usersData.adminId), {
        [usersData.roomChatId + USER_CHATS_FIELD.USER_INFO]: {
          uid: usersData.adminId,
          displayName: usersData.displayName,
          photoURL: usersData.avatarUrl,
        },
        [usersData.roomChatId + USER_CHATS_FIELD.DATE]: serverTimestamp(),
      });

      await updateDoc(doc(db, FIREBASE_CHAT.USER_CHATS, usersData.userId), {
        [usersData.roomChatId + USER_CHATS_FIELD.USER_INFO]: {
          uid: usersData.userId,
          displayName: usersData.displayName,
          photoURL: usersData.avatarUrl,
        },
        [usersData.roomChatId + USER_CHATS_FIELD.DATE]: serverTimestamp(),
      });
    }
  }, [user]);

  const fetchData = async () => {
    // Get user data
    const usersData = await getCurrentUser(user);

    // Check if usersData is undefined before accessing its properties
    if (usersData) {
      const res = await getDoc(
        doc(db, FIREBASE_CHAT.CHATS, usersData.roomChatId),
      );

      setUserChat(usersData);
      if (res.exists()) {
        setMessages(res.data().messages);
      } else {
        await createChatRoom();
      }
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  useEffect(() => {
    if (!userChat.roomChatId) return;
    const unSub = onSnapshot(
      doc(db, FIREBASE_CHAT.CHATS, userChat.roomChatId),
      (doc) => {
        doc.exists() && setMessages(doc.data().messages);
      },
    );

    if (boxRef.current) {
      boxRef.current.scrollTop = boxRef.current.scrollHeight;
    }

    return () => {
      unSub();
    };
  }, [userChat.roomChatId]);

  return (
    <Box w="full" bg="background.body.quaternary" borderRadius="lg">
      <Flex
        direction="row"
        justifyContent="space-between"
        alignItems="center"
        justify="center"
        borderBottom="1px solid"
        borderBottomColor="border.tertiary"
        padding="24px 26px"
      >
        <Heading
          as="h3"
          fontWeight="semibold"
          color="text.primary"
          fontSize="2xl"
          textTransform="capitalize"
        >
          team chat
        </Heading>
      </Flex>

      <Box padding={{ base: '24px 20px', lg: '45px 35px' }}>
        <Box
          ref={boxRef}
          overflowX="auto"
          overflowY="scroll"
          css={{
            '&::-webkit-scrollbar': {
              width: 2,
            },
            '&::-webkit-scrollbar-track': {
              width: 2,
            },
            '&::-webkit-scrollbar-thumb': {
              background: 'gray',
              borderRadius: '24px',
            },
          }}
          maxHeight={361}
          padding={5}
        >
          {messages.map((m) => (
            <Message
              content={m.text}
              key={m.date.second}
              senderId={m.senderId}
              avatarAdmin={userChat.avatarAdminUrl}
              avatarUser={userChat.avatarUrl}
            />
          ))}
        </Box>
        <InputSendMessages boxRef={boxRef} />
      </Box>
    </Box>
  );
};

const BoxChat = memo(BoxChatComponent);

export default BoxChat;
