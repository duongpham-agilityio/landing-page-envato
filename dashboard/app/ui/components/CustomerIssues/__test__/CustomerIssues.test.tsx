import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import { CustomerIssues } from '@/ui/components';

const fetchNextPage = jest.fn();
jest.mock('@/lib/hooks', () => ({
  useGetListIssues: () => ({
    fetchNextPage,
    hasNextPage: true,
    isLoading: false,
    isFetchingNextPage: false,
  }),
}));

describe('CustomerIssues render', () => {
  test('Should render match with snapshot.', () => {
    const { container } = render(<CustomerIssues onLoadMore={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });

  test('calls fetchNextPage on button click', async () => {
    const { getByText } = render(<CustomerIssues onLoadMore={fetchNextPage} />);

    const showMoreButton = getByText('Load More');
    await userEvent.click(showMoreButton);

    await waitFor(() => expect(fetchNextPage).toHaveBeenCalled());
  });
});
