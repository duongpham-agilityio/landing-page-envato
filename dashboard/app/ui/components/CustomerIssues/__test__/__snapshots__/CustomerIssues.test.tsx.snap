// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerIssues render Should render match with snapshot. 1`] = `
<div>
  <div
    class="css-1rdd54d"
  >
    <p
      class="chakra-text css-vk907l"
    >
      Recent Support
    </p>
    <div
      class="css-k008qs"
    >
      <div
        class="css-1ten0p5"
      >
        <img
          alt="This is avatar"
          class="chakra-image css-xerxct"
          src="/images/avatar.webp"
        />
        <div
          class="css-m75dnw"
        >
          <div
            class="css-k008qs"
          >
            <p
              class="chakra-text css-3ptjhf"
            >
              Joel Becker
              <span
                class="chakra-text css-1y8hbw7"
              >
                12h
              </span>
            </p>
          </div>
          <p
            class="chakra-text css-31rn8f"
          >
            When is the Sketch version available for download?
          </p>
          <p
            class="chakra-text css-qnadlj"
          >
            Love this so much! What tools do you use to create your 3d illustrations ?
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-k008qs"
    >
      <div
        class="css-1ten0p5"
      >
        <img
          alt="This is the big avatar"
          class="chakra-image css-xerxct"
          src="/images/big-avatar.webp"
        />
        <div
          class="css-m75dnw"
        >
          <div
            class="css-k008qs"
          >
            <p
              class="chakra-text css-3ptjhf"
            >
              Joel Becker
              <span
                class="chakra-text css-1y8hbw7"
              >
                1h
              </span>
            </p>
          </div>
          <p
            class="chakra-text css-31rn8f"
          >
            When is the Sketch version available for download?
          </p>
          <p
            class="chakra-text css-qnadlj"
          >
            Love this so much! What tools do you use to create your 3d illustrations ?
          </p>
        </div>
      </div>
    </div>
    <div
      class="css-k008qs"
    >
      <div
        class="css-1ten0p5"
      >
        <img
          alt="This is avatar sign up"
          class="chakra-image css-xerxct"
          src="/images/avatar-sign-up.webp"
        />
        <div
          class="css-m75dnw"
        >
          <div
            class="css-k008qs"
          >
            <p
              class="chakra-text css-3ptjhf"
            >
              Joel Becker
              <span
                class="chakra-text css-1y8hbw7"
              >
                6h
              </span>
            </p>
          </div>
          <p
            class="chakra-text css-31rn8f"
          >
            When is the Sketch version available for download?
          </p>
          <p
            class="chakra-text css-qnadlj"
          >
            Love this so much! What tools do you use to create your 3d illustrations ?
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
